### PR TITLE
Implement automatic release script

### PR DESCRIPTION
**Pull request by jomrazek**

### CODE REVIEW

File: release.sh [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+
+echo "This is release script for pnc-api-types-ts repo.
+
+usage: './release.sh pnc-revision' where pnc revision is tag or commit hash
+
+script will generate new api types, change version and commit changes automatically.
+You can then check the result and push it. 
+"
+
+if [[ -z ${1} ]]
```
Comment (janinko) [inline release.sh]:
```suggestion
if [[ -z "${1}" ]]
```

File: release.sh [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+
+cd pnc && git reset --hard ${PNC_REV}
+
+cd rest-api
+
+mvn clean install -DskipTests
+
+MVN_PNC_VERSION=$(mvn -q -Dexec.executable=echo -Dexec.args='${project.version}' --non-recursive exec:exec)
+
+cd .. && cd ..
```
Comment (matedo1) [inline release.sh]: maybe just `cd ../..` could be sufficient ?

File: release.sh [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+
+PNC_REV=${1}
+PNC_VER=${PNC_REV}
+if [ ${#PNC_REV} -eq 40 ]; then
+   PNC_VER=$(git rev-parse --short ${PNC_REV});
+fi
+
+mkdir .tmp
+
+cd .tmp
```
Comment (janinko) [inline release.sh]: I prefer `pushd` and `popd` instead of `cd` and `cd ..` in scripts

File: release.sh [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+
+wget https://repo1.maven.org/maven2/org/jboss/pnc/project-manipulator/project-manipulator-cli/1.1.0/project-manipulator-cli-1.1.0.jar
+
+git clone https://github.com/project-ncl/pnc.git
+
+cd pnc && git reset --hard ${PNC_REV}
```
Comment (janinko) [inline release.sh]: why is the git reset conditioned on `cd pnc` succeeding?

File: release.sh [hunk 5]
```diff
@@ -0,0 +1,62 @@
+#!/bin/sh
+
+echo "This is release script for pnc-api-types-ts repo.
+
+usage: './release.sh pnc-revision' where pnc revision is tag or commit hash
+
+script will generate new api types, change version and commit changes automatically.
+You can then check the result and push it. 
+"
```
Comment (janinko) [inline release.sh:10]: consider adding
```
# exit when any command fails
set -e
```

File: release.sh [hunk 6]
```diff
@@ -0,0 +1,62 @@
+#!/bin/sh
+
+echo "This is release script for pnc-api-types-ts repo.
+
+usage: './release.sh pnc-revision' where pnc revision is tag or commit hash
```
Comment (matedo1) [inline release.sh:5]: maybe some realworld examples could be provided?

File: release.sh [hunk 7]
```diff
@@ -0,0 +1,62 @@
...
+
+cd ..
+
+java -jar .tmp/project-manipulator-cli-1.1.0.jar -f package.json -DversionOverride=${NPM_VERSION}
+
+npm run build
+
+rm -rf .tmp
+
+git add *
```
Comment (janinko) [inline release.sh]: `git add *` ignores files starting with `.` in the current directory. You can use `git add .` to include everything in current directory and subdirectories

**Review by janinko — CHANGES_REQUESTED** [review]

File: release.sh [hunk 8]
```diff
@@ -0,0 +1,62 @@
...
+
+git clone https://github.com/project-ncl/pnc.git
+
+cd pnc && git reset --hard ${PNC_REV}
+
+cd rest-api
+
+mvn clean install -DskipTests
+
+MVN_PNC_VERSION=$(mvn -q -Dexec.executable=echo -Dexec.args='${project.version}' --non-recursive exec:exec)
```
Comment (matedo1) [inline release.sh:42]: maybe some comment containing real example, like `2.4.0-SNAPSHOT`

File: release.sh [hunk 9]
```diff
@@ -0,0 +1,62 @@
...
+
+cp pnc/rest-api/target/pnc-openapi.json .
+
+NPM_VERSION=(v${MVN_PNC_VERSION}-pnc-${PNC_VER})
+
+echo $NPM_VERSION
+
+cd ..
+
+java -jar .tmp/project-manipulator-cli-1.1.0.jar -f package.json -DversionOverride=${NPM_VERSION}
```
Comment (matedo1) [inline release.sh:57]: please add some comment

File: release.sh [hunk 10]
```diff
@@ -0,0 +1,62 @@
...
+
+mvn clean install -DskipTests
+
+MVN_PNC_VERSION=$(mvn -q -Dexec.executable=echo -Dexec.args='${project.version}' --non-recursive exec:exec)
+
+cd .. && cd ..
+
+cp pnc/rest-api/target/pnc-openapi.json .
+
+NPM_VERSION=(v${MVN_PNC_VERSION}-pnc-${PNC_VER})
```
Comment (matedo1) [inline release.sh]: please add some comment containing real example

File: release.sh (hunk 7 above)
Comment (jomrazek) [reply release.sh]: do I want that? it should generally commit just package.json , package-lock.json and index.dt.s

**Comment by matedo1** [comment]
Thanks @jomrazek for this PR! According to our discussion you could also update README file to contain possible scenarios:

1a: jenkins - fully automated for each PNC release (preffered)
1b: jenkins - manually executed for custom PNC tag / hash (preffered when custom, not released, PNC revision is needed)
2: manual release process without jenkins (deprected, only if jenkins job is broken)

**Comment by jomrazek** [comment]
> Thanks @jomrazek for this PR! According to our discussion you could also update README file to contain possible scenarios:
> 
> 1a: jenkins - fully automated for each PNC release (preffered) 1b: jenkins - manually executed for custom PNC tag / hash (preffered when custom, not released, PNC revision is needed) 2: manual release process without jenkins (deprected, only if jenkins job is broken)

@matedo1 should I also mention 2a: old process, 2b: running release.sh locally and just pushing the result?

**Comment by matedo1** [comment]
@jomrazek yes, you can mention both 2a and 2b, and in the future, once your script is well tested, 2a would be removed

File: README.md [hunk 11]
```diff
@@ -4,7 +4,27 @@ TypeScript based types used in PNC project.
 
 This project clones **PNC project** and builds `pnc/rest-api` module to get `pnc-openapi.json` specification using `generate-openapi-spec.sh` script and transform it to `index.d.ts` using `dtsgen` command.
 
-## Generate new types and release new version
+## Releasing new version of pnc-api-types-ts
+
+### Remote & automatic
```
Comment (matedo1) [inline README.md]: I would mark it as a preferred way `### Remote & automatic (preferred)`

File: README.md [hunk 12]
```diff
@@ -4,7 +4,27 @@ TypeScript based types used in PNC project.
...
+
+#### Option A - fully automatic
+
+New pnc-api-types-ts release is created by our jenkins instance automatically when there is new tag created in **PNC** project repository.
+
+#### Option B - one user click automatic
+
+When you need to create new pnc-api-types-ts release from unreleased PNC revision you can go to our jenkins instance and run `pnc-api-types-release` job. Job will ask for **PNC** revision to create new release from and takes care of rest.
+
+### Local
```
Comment (matedo1) [inline README.md]: I would mark it as a deprecated way `### Local (deprecated, only if Remote is not working)`

File: release.sh (hunk 7 above)
Comment (matedo1) [reply release.sh]: using `git add .` could be potentially less error prone in the future when also some other files starting with `.` would have to be committed, I would prefer `git add *` only if it's intentional not to commit files starting with `.`

**Review by janinko — APPROVED** [review]

File: release.sh (hunk 6 above)
Comment (jomrazek) [reply release.sh:5]: is this old comment you forgot to add yesterday when we talked about it?

File: README.md (hunk 11 above)
Comment (jomrazek) [reply README.md]: (it's not working yet)

File: release.sh (hunk 7 above)
Comment (jomrazek) [reply release.sh]: I don't mind changing it, but it was intentional as it should generally add just files I mentioned

File: release.sh (hunk 6 above)
Comment (matedo1) [reply release.sh:5]: yes, it seems that they were not submitted successfully, today I added some additional comment so they were submitted together. I will resolve old ones that were solved already and keep the ones that are still valid.

**Review by matedo1 — APPROVED** [review]

**Pull request merged**